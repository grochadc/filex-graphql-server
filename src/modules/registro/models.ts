import internal from "stream";

export { ServerContext } from "../../server";
type Nivel = "1" | "2" | "3" | "4" | "5" | "6";
type Curso = "en" | "fr";
type Genero = "M" | "F";
type Url = string;

export type ApplicantModel = {
  codigo: string;
  nombre: string;
  apellido_paterno: string;
  apellido_materno: string;
  genero: Genero;
  ciclo: string;
  carrera: string;
  telefono: string;
  email: string;
  externo: boolean;
  desertor: boolean;
  nivel: Nivel;
  curso: Curso;
};

export type ScheduleModel = {
  id: number;
  ciclo: string;
  name: string;
  time: string;
  aula: string;
  teacher_id: number;
  nivel: number;
  teacher: string;
};

type RgisteringLevelsEndpoint = {
  en: Nivel[];
  fr: Nivel[];
};

type LevelsString =
  | "level1"
  | "level2"
  | "level3"
  | "level4"
  | "level5"
  | "level6";

export type SchedulesEndpoint = {
  //DB -> prod/scheduels.json
  en: {
    [key in LevelsString]: {
      [group: string]: ScheduleModel;
    };
  };
  fr?: {
    [key in LevelsString]?: {
      [group: string]: ScheduleModel;
    };
  };
};

type AlreadyRegisteredEndpoint = {
  [codigo: string]: string; //"E4-1"
};
