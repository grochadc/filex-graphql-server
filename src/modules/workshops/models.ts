export type DatabaseModel = {
  workshops: {
    [workshop_id: string]: WorkshopModel;
  };
  options: {
    [option_id: string]: OptionModel;
  };
  teachers: {
    [teacher_id: string]: TeacherModel;
  };
  studentsReservations?: {
    [codigo: string]: StudentReservationModel;
  };
  availableOptions?: {
    [option_id: string]: number;
  };
  system: {
    max_reservations: number;
  };
};

export type StudentsDBModel = {
  [codigo: string]: StudentModel;
};

export type WorkshopModel = {
  id: string;
  name: string;
  description: string;
  levels: string[];
  option_ids: string[];
};

export interface OptionModel {
  id: string;
  day: string;
  time: string;
  teacher_id: string;
  teacher_name: string;
  workshop_id: string;
  workshop_name: string;
  isTutorial: boolean;
  available: boolean;
  active: boolean;
  url: string;
}

export type TeacherModel = {
  id: string;
  name: string;
  option_ids: string[];
  raw_reservations?: {
    [option_id: string]: {
      [random: string]: TeacherReservationModel;
    };
  };
};

export type TeacherReservationModel = {
  codigo: string;
  nombre: string;
  apellido_paterno: string;
  apellido_materno: string;
  nivel: string;
  grupo: string;
  workshop_id: string;
  workshop_name: string;
  option_id: string;
  option_name: string;
  tutorial_reason?: string;
};

export type StudentReservationModel = {
  option_id: string;
};

export type StudentModel = {
  id: number;
  codigo: string;
  nombre: string;
  apellido_paterno: string;
  apellido_materno: string;
  genero: string;
  carrera: string;
  ciclo: string;
  telefono: string;
  email: string;
  nivel: string;
  curso: string;
  externo: boolean;
  grupo: string;
};
