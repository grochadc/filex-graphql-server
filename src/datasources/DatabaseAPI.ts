import { ApolloError } from "apollo-server";
import { DataSource } from "apollo-datasource";
import { Workshop, Teacher, TeacherOption } from "../generated/graphql";
import { ParameterizedQuery as PQ } from "pg-promise";

export interface OptionModel {
  id: number;
  day: string;
  time: string;
  teacher_name: string;
  url: string;
  workshop_name: string;
  workshop_description: string;
  workshop_levels: number[];
  teacher_id: number;
  workshop_id: number;
  reservations: number;
}

export type StudentModel = {
  id: number;
  codigo: string;
  nombre: string;
  apellido_paterno: string;
  apellido_materno: string;
  genero: string;
  carrera: string;
  ciclo: string;
  telefono: string;
  email: string;
  nivel: string;
  curso: string;
  externo: boolean;
  grupo: string;
};

type Option = Omit<
  OptionModel,
  "id" | "workshop_id" | "teacher_id" | "reservations"
> & {
  id: string;
  workshop_id: string;
  teacher_id: string;
  isTutorial: boolean;
  available: boolean;
};

export const SELECT_STUDENT = `SELECT * FROM student WHERE codigo=$1`;

export const SELECT_STUDENT_RESERVATION = `
SELECT
  r.*,
  o.day,
  o.time,
  o.url,
  o.teacher_id,
  t.name as teacher_name,
  o.workshop_id,
  w.name as workshop_name
FROM reservation r
INNER JOIN option o ON o.id=r.option_id
INNER JOIN teacher t ON t.id=o.teacher_id
INNER JOIN workshop w ON w.id=o.workshop_id
WHERE r.student_id=$1
`;

export const SELECT_NEW_RESERVATION = `
SELECT
  r.id,
  o.day,
  o.time,
  t.name as teacher_name,
  t.id as teacher_id,
  w.name as workshop_name,
  w.id as workshop_id,
  o.url
FROM reservation r
INNER JOIN option o ON r.option_id=o.id
INNER JOIN teacher t ON o.teacher_id=t.id
INNER JOIN workshop w ON o.workshop_id=w.id
WHERE r.id=$1
`;

export const SELECT_OPTIONS = `
SELECT
  Option.id,
  Workshop.name as workshop_name,
  Option.day,
  Option.time,
  Teacher.name as teacher_name,
  Option.url,
  Workshop.description as workshop_description,
  Workshop.levels as workshop_levels,
  Teacher.id as teacher_id,
  Workshop.id as workshop_id,
  COUNT(Reservation.id) as reservations
FROM Option
INNER JOIN Teacher ON Option.teacher_id=Teacher.id
INNER JOIN Workshop on Option.workshop_id=Workshop.id
LEFT JOIN Reservation ON Option.id=Reservation.option_id
GROUP BY Option.id, Workshop.id, Teacher.id
ORDER BY Option.id;
`;

export const SELECT_TEACHER_INFO = `
SELECT
  id,
  name,
  json_agg(
        (
        SELECT
          json_build_object('option_id',r.option_id, 'students', json_agg(s))
        FROM reservation r
        INNER JOIN student s ON s.id=r.student_id
        WHERE r.option_id=any(teacher.option_ids)
        GROUP BY r.option_id
        )
  ) as options FROM teacher Where id=$1 GROUP BY teacher.id;
`;

export type TeacherReservation = {
  reservation_id: number;
  codigo: string;
  nombre: string;
  apellido_materno: string;
  apellido_paterno: string;
  teacher_id: number;
  teacher: string;
  workshop: string;
  time: string;
  day: string;
  option_id: string;
};
export const SELECT_TEACHER_RESERVATIONS = `
SELECT
  t.id as teacher_id,
  r.id as reservation_id,
  codigo,
  nombre,
  apellido_paterno,
  apellido_materno,
  t.id as teacher_id,
  t.name as teacher,
  w.name as workshop,
  o.time,
  o.day,
  o.url,
  o.id as option_id
FROM teacher t
INNER JOIN option o ON o.teacher_id=t.id
INNER JOIN workshop w ON o.workshop_id=w.id
INNER JOIN reservation r ON r.option_id=o.id
INNER JOIN student s ON s.id=r.student_id
WHERE t.id=$1
`;

export const SELECT_SINGLE_OPTION = `
SELECT * FROM Option o WHERE o.id=$1;
`;

class DatabaseAPI extends DataSource {
  db: any;
  constructor(db: any) {
    super();
    console.log("construting databaseAPI");
    this.db = db;
  }
  _getStudentSql(query: string, values: any[]) {
    const paramQuery = new PQ({ text: query, values });
    return this.db.any(paramQuery);
  }
  async getStudent(codigo: string) {
    if (/^[1234567890]+$/.test(codigo)) {
      const result: StudentModel[] = await this._getStudentSql(SELECT_STUDENT, [
        codigo
      ]);
      if (result.length === 0)
        throw new ApolloError("404: Alumno inexistente", "404");
      return result[0];
    } else {
      throw new ApolloError("Input wasn't a number");
    }
  }
  async getOptionsByIDs(ids: string[]): Promise<Option> {
    const parsedIDs = ids.map(id => Number(id));
    const result = await this.db.any(
      new PQ({ text: SELECT_SINGLE_OPTION, values: [parsedIDs] })
    );
    return result[0];
  }
  async getAllOptions(max_reservations: number): Promise<Option[]> {
    const options: OptionModel[] = await this.db.any(SELECT_OPTIONS);
    const mapped = options.map(option => {
      return {
        ...option,
        available: Boolean(Number(option.reservations) < max_reservations),
        isTutorial: Boolean(option.workshop_id === 4),
        id: option.id.toString(),
        workshop_id: option.workshop_id.toString(),
        teacher_id: option.teacher_id.toString()
      };
    });
    return mapped;
  }

  async getAllWorkshops(max_reservations: number): Promise<Workshop[]> {
    const options = await this.getAllOptions(max_reservations);
    const workshops: Workshop[] = [];
    options.forEach(option => {
      const { workshop_description, ...simpleOption } = option;
      const foundWorkshopIndex = workshops.findIndex(
        workshop => workshop.id === option.workshop_id
      );
      if (foundWorkshopIndex === -1) {
        const new_workshop = {
          id: option.workshop_id,
          name: option.workshop_name,
          description: option.workshop_description,
          levels: option.workshop_levels.map(level => level.toString()),
          options: [simpleOption]
        };
        workshops.push(new_workshop);
      } else {
        workshops[foundWorkshopIndex].options.push(simpleOption);
      }
    });
    return workshops;
  }
  _getStudentReservationSql(query: string, values: any[]) {
    const paramQuery = new PQ({ text: query, values });
    return this.db.any(paramQuery);
  }
  getStudentReservation(id: number) {
    return this._getStudentReservationSql(SELECT_STUDENT_RESERVATION, [id]);
  }
  async _makeReservation(query: string, values: any[]) {
    const paramQuery = new PQ({ text: query, values });
    const result = await this.db.any(paramQuery);
    return result;
  }
  _newReservation(id: number) {
    return this.db.any(new PQ({ text: SELECT_NEW_RESERVATION, values: [id] }));
  }
  async makeReservation(
    student_id: number,
    option_id: number,
    tutorial_reason?: string
  ) {
    const INSERT_RESERVATION = `
      INSERT INTO reservation (student_id, option_id, tutorial_reason) VALUES ($1,$2,$3) RETURNING id;
    `;
    const normalizedTutorialReason =
      tutorial_reason === undefined ? null : tutorial_reason;
    const reservations = await this._makeReservation(INSERT_RESERVATION, [
      student_id,
      option_id,
      normalizedTutorialReason
    ]);
    const result = await this._newReservation(reservations[0].id);
    return result;
  }
  _resetReservations(query) {
    return this.db.any(query);
  }
  async resetReservations() {
    const DELETE_RESERVATIONS = `DELETE FROM reservation`;
    try {
      this._resetReservations(DELETE_RESERVATIONS);
    } catch (e) {
      throw new ApolloError(e);
    }
    return true;
  }
  _getTeacher(id: number) {
    const parsedQuery = new PQ({
      text: SIMPLE_TEACHER_INFO,
      values: [id]
    });
    return this.db.any(parsedQuery);
  }
  async getTeacher(id: string) {
    const parsedID = Number(id);
    const result: SimpleTeacherInfo[] = await this._getTeacher(parsedID);
    return {
      id: result[0].id.toString(),
      name: result[0].name,
      option_ids: result[0].option_ids.map(id => id.toString()),
      options: result.map(teacher => {
        return {
          id: teacher.option_id.toString(),
          day: teacher.day,
          time: teacher.time,
          teacher_name: teacher.name,
          teacher_id: teacher.id.toString(),
          workshop_name: teacher.workshop_name,
          workshop_id: teacher.workshop_id.toString(),
          url: teacher.url
        };
      })
    };
  }
  async getAllTeachers() {
    const result: SelectTeachers = await this.db.any(SELECT_TEACHERS);
    return result.map(teacher => {
      return {
        ...teacher,
        id: teacher.id.toString(),
        option_ids: teacher.option_ids.map(id => id.toString())
      };
    });
  }
  _getTeacherReservations(option_id: number) {
    return this.db.any(
      new PQ({ text: GET_TEACHER_RESERVATIONS, values: [option_id] })
    );
  }
  getTeacherReservations(option_id: string) {
    const parsedOptionID = Number(option_id);
    return this._getTeacherReservations(parsedOptionID);
  }
  _deleteOptionReservations(option_id: number) {
    return this.db.any(
      new PQ({ text: DELETE_TEACHER_RESERVATIONS, values: [option_id] })
    );
  }
  deleteOptionReservations(option_id: string) {
    const parsedID = Number(option_id);
    this._deleteOptionReservations(parsedID);
  }
}

export const GET_TEACHER_RESERVATIONS = `
SELECT
  r.id,
  w.id as workshop_id,
  w.name as workshop_name,
  o.id as option_id,
  s.codigo,
  s.nombre,
  s.apellido_paterno,
  s.apellido_materno,
  s.nivel,
  s.grupo,
  r.tutorial_reason
FROM reservation r
INNER JOIN option o ON r.option_id=o.id
INNER JOIN workshop w ON o.workshop_id=w.id
INNER JOIN student s ON s.id=r.student_id
WHERE o.id=$1
`;
export type SelectTeachers = {
  id: number;
  name: string;
  option_ids: number[];
}[];
export const SELECT_TEACHERS = `
SELECT * FROM Teacher;
`;

export type SimpleTeacherInfo = {
  id: number;
  name: string;
  workshop_name: string;
  workshop_id: number;
  option_id: number;
  day: string;
  time: string;
  url: string;
  option_ids: number[];
};
export const SIMPLE_TEACHER_INFO = `
SELECT t.id, t.name, w.name as workshop_name, w.id as workshop_id, o.id as option_id, o.day, o.time, o.url, t.option_ids
FROM teacher t
INNER JOIN option o ON o.teacher_id=t.id
INNER JOIN workshop w ON w.id=o.workshop_id
WHERE t.id=$1
`;

export const DELETE_TEACHER_RESERVATIONS = `
  DELETE FROM reservation WHERE option_id=$1
`;

export default DatabaseAPI;
