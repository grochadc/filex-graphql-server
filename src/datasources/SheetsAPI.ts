import { DataSource } from "apollo-datasource";
import { google } from "googleapis";
require("dotenv").config();

function getJWT() {
  return new google.auth.JWT(
    process.env.GOOGLE_CLIENT_EMAIL,
    null,
    process.env.GOOGLE_PRIVATE_KEY.replace(/\\n/g, "\n"),
    ["https://www.googleapis.com/auth/spreadsheets"]
  );
}

class SheetsAPI extends DataSource {
  sheetID: any;
  api: any;
  ready: boolean;
  constructor(sheetID) {
    super();
    this.sheetID = sheetID;
    this.api = google.sheets({ version: "v4", auth: getJWT() });
  }

  async saveApplicant(applicant: Applicant) {
    const values = [
      [
        new Date().toUTCString(),
        applicant.id,
        applicant.codigo,
        applicant.nombre,
        applicant.apellido_paterno,
        applicant.apellido_materno,
        applicant.genero,
        applicant.ciclo,
        applicant.carrera,
        applicant.telefono,
        applicant.email,
        applicant.externo,
        applicant.reubicacion,
        applicant.curso,
        applicant.meetLink,
        applicant.nivel_escrito,
      ],
    ];
    this.api.spreadsheets.values.append({
      spreadsheetId: this.sheetID,
      range: "Today!A1",
      valueInputOption: "RAW",
      resource: {
        values,
      },
    });
  }
}

export { SheetsAPI };
