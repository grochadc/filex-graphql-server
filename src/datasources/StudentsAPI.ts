import { RESTDataSource } from "apollo-datasource-rest";

export type GetStudentFn = (codigo: string, env?: 'dev'|'prod') => Promise<any>;

class StudentsAPI extends RESTDataSource {
  constructor() {
    super();
    this.baseURL = "https://filex-5726c.firebaseio.com/students";
  }

  async _get(url: string, env:string|undefined) {
    const finalUrl = env ? '/'+env+url : '/prod'+url;
    return this.get(finalUrl);
  }

  async getStudent(...[codigo, env]: Parameters<GetStudentFn>) {
    return this.get(`/${env}/${codigo}.json`);
  }
}

export { StudentsAPI };
