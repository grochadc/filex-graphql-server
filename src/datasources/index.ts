import { FirebaseAPI } from "./myREST";
import { RegistroAPI } from "./RegistroAPI";
import { ExamAPI } from "./ExamAPI";
import { WorkshopsAPI } from "./WorkshopsAPI";
import { StudentsAPI } from "./StudentsAPI";

export { RegistroAPI, FirebaseAPI, ExamAPI, WorkshopsAPI, StudentsAPI };
