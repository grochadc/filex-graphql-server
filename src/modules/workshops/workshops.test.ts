import testServer from "testUtils/testServer";
import { gql } from "apollo-server";
import mocks from "../../datasources/DatabaseAPI/mocks";
import DatabaseAPI from "datasources/DatabaseAPI";
import studentMocks from "datasources/StudentsAPI/mocks";
import { StudentsAPI } from "datasources/StudentsAPI";
import Database from "testUtils/databaseCreator";
import { WorkshopsAPI } from "datasources/WorkshopsAPI";
import { SheetsAPI } from "datasources/SheetsAPI";
import { MutationSaveWorkshopsAttendanceArgs } from "generated/graphql";
import { DatabaseModel, StudentsDBModel } from "./models";

const databaseAPI = new DatabaseAPI(mocks.db);

type OptionalDatabaseModel = Partial<DatabaseModel>;
const db: { prod: OptionalDatabaseModel } = {
  prod: {
    system: {
      max_reservations: 30
    },
    workshops: {
      workshop_id1: {
        id: "workshop_id1",
        name: "Conversation",
        description: "This is the conversation workshop description.",
        levels: ["1", "2", "3", "4", "5", "6"],
        option_ids: ["option_id1", "option_id2"]
      }
    },
    teachers: {
      teacher_id1: {
        id: "teacher_id1",
        name: "Fulanito",
        option_ids: ["option_id1", "option_id2"],
        raw_reservations: {
          option_id1: {
            _randomstring1: {
              codigo: "1234567890",
              nombre: "Benito Antonio",
              apellido_paterno: "Martinez",
              apellido_materno: "Ocasio",
              nivel: "4",
              grupo: "E4-1",
              workshop_id: "workshop_id1",
              workshop_name: "Conversation",
              option_id: "option_id1",
              option_name: "why?"
            }
          }
        }
      },
      teacher_id2: {
        id: "teacher_id2",
        name: "Manganito",
        option_ids: ["option_id1"]
      }
    },
    options: {
      option_id1: {
        active: true,
        available: true,
        id: "id1",
        day: "Lunes",
        time: "1300 - 14:00",
        teacher_id: "teacher_id1",
        teacher_name: "Fulanito",
        workshop_id: "workshop_id1",
        workshop_name: "Conversation",
        url: "http://url1",
        isTutorial: false
      },
      option_id2: {
        id: "id2",
        active: true,
        available: true,
        day: "Martes",
        time: "13:00 - 14:00",
        teacher_id: "teacher_id1",
        teacher_name: "Fulanito",
        workshop_id: "workshop_id1",
        workshop_name: "Conversation",
        url: "http://url2",
        isTutorial: false
      }
    },
    availableOptions: {
      option_id1: 10
    },
    studentsReservations: {
      "1234509876": {
        option_id: "option_id1"
      }
    }
  }
};

const studentsDB = {
  prod: {
    "1234567890": {
      codigo: "1234567890",
      nombre: "Benito Antonio",
      apellido_paterno: "Martinez",
      apellido_materno: "Ocasio",
      genero: "M",
      carrera: "Abogado",
      ciclo: "2021A",
      telefono: "1234567890",
      email: "bad@bunny.pr",
      nivel: "4",
      curso: "en",
      grupo: "E4-1",
      externo: false
    },
    "1234509876": {
      codigo: "1234509876",
      nombre: "Juan",
      apellido_materno: "Preciado",
      apellido_paterno: "Paramo",
      genero: "M",
      carrera: "Abogado",
      ciclo: "2021A",
      telefono: "1234567890",
      email: "juan@lamedialuna.com",
      nivel: "5",
      curso: "en",
      grupo: "E5-1",
      externo: false
    }
  }
};
const database = new Database(db);
const workshopsAPI = new WorkshopsAPI();
workshopsAPI.isOpen = jest.fn(() => Promise.resolve(true));
workshopsAPI.getMaxStudentReservations = jest.fn(() => Promise.resolve(30));

const studentsDatabase = new Database(studentsDB);
const studentsAPI = new StudentsAPI(studentMocks.db);

const workshopsSheetsAPI = new SheetsAPI("sheetsID");
workshopsSheetsAPI.append = jest.fn(() => Promise.resolve());

const dataSources = () => {
  return {
    workshopsAPI: workshopsAPI,
    studentsAPI: studentsAPI,
    workshopsSheetsAPI: workshopsSheetsAPI,
    databaseAPI: databaseAPI
  };
};

const context = () => {
  return {
    enviroment: "prod"
  };
};
const { query } = testServer(dataSources, context);

afterEach(() => {
  databaseAPI.db.one.mockClear();
});

test("gets all workshops", async () => {
  const GET_WORKSHOPS = gql`
    query getWorkshops {
      workshops {
        name
        description
        levels
        options {
          id
          day
          time
          teacher_name
          workshop_name
          url
          zoom_id
          available
        }
      }
    }
  `;
  const res = await query({
    query: GET_WORKSHOPS
  });
  expect(res.errors).toBeUndefined();
  expect(res.data).toMatchSnapshot();
});

describe("reservations", () => {
  const MAKE_RESERVATION = gql`
    mutation makeReservation(
      $student_id: ID!
      $option_id: ID!
      $tutorial_reason: String
    ) {
      makeWorkshopReservation(
        student_id: $student_id
        option_id: $option_id
        tutorial_reason: $tutorial_reason
      ) {
        id
        day
        time
        teacher_id
        teacher_name
        workshop_name
        workshop_id
        url
        zoom_id
      }
    }
  `;
  test("makes a reservation", async () => {
    const variables = {
      student_id: "1",
      option_id: "1"
    };
    const res = await query({ query: MAKE_RESERVATION, variables });
    expect(res.errors).toBeUndefined();
    expect(databaseAPI.db.one.mock.calls[0][0].values).toMatchSnapshot();
    expect(res.data).toMatchSnapshot();
  });

  test("makes a tutorials reservation", async () => {
    const variables = {
      student_id: "1",
      option_id: "1",
      tutorial_reason: "negative questions"
    };
    const res = await query({ query: MAKE_RESERVATION, variables });
    expect(res.errors).toBeUndefined();
    expect(databaseAPI.db.one.mock.calls[0][0].values).toMatchSnapshot();
    expect(res.data).toMatchSnapshot();
  });

  test("resets reservations", async () => {
    const RESET_RESERVATIONS = gql`
      mutation reset {
        resetReservations
      }
    `;
    const res = await query({ query: RESET_RESERVATIONS });
    expect(res.errors).toBeUndefined();
    expect(res.data.resetReservations).toBe(true);
  });

  test("errors when student already has a reservation", async () => {
    const variables = {
      student_id: "2",
      option_id: "2"
    };
    const res = await query({ query: MAKE_RESERVATION, variables });
    expect(res.errors[0].extensions.code).toBe("RESERVATION_FORBIDDEN");
  });
});

describe("student data", () => {
  const GET_STUDENT = gql`
    query getStudent($codigo: ID!) {
      student(codigo: $codigo) {
        nombre
        reservation {
          day
          time
          teacher_name
          workshop_name
          url
        }
      }
    }
  `;

  test("normal not registered", async () => {
    const variables = { codigo: "1234567890" };
    const res = await query({ query: GET_STUDENT, variables });
    expect(res.errors).toBeUndefined();
    expect(res.data.student.reservation).toBeNull();
    expect(res.data).toMatchSnapshot();
  });
  test("already registered", async () => {
    const variables = { codigo: "0987654321" };
    const res = await query({ query: GET_STUDENT, variables });
    expect(res.errors).toBeUndefined();
    expect(res.data.student.reservation).toBeTruthy();
    expect(res.data).toMatchSnapshot();
  });
  test("not found", async () => {
    const variables = { codigo: "1234509876" };
    const res = await query({ query: GET_STUDENT, variables });
    expect(res.errors[0].message).toBe("404: Alumno inexistente");
  });
});

describe("teacher dashboard", () => {
  const GET_TEACHER_DASHBOARD = gql`
    query getTeacherDashboard($teacher_id: ID!) {
      teacher(id: $teacher_id) {
        id
        name
        options {
          id
          day
          time
          url
          reservations {
            codigo
            nombre
            apellido_paterno
            apellido_materno
            nivel
            grupo
          }
        }
      }
    }
  `;

  test("normal", async () => {
    const variables = {
      teacher_id: "1"
    };
    const res = await query({ query: GET_TEACHER_DASHBOARD, variables });
    expect(res.errors).toBeUndefined();
    expect(res.data).toMatchSnapshot();
  });
  test("no reservations", async () => {
    const variables = { teacher_id: "2" };
    const res = await query({ query: GET_TEACHER_DASHBOARD, variables });
    if (res.errors) console.log(JSON.stringify(res.errors));
    expect(res.errors).toBeUndefined();
    expect(res.data).toMatchSnapshot();
  });

  test("set workshop link", async () => {
    const SET_WORKSHOP_LINK = gql`
      mutation setLink($option_id: ID!, $url: String!) {
        setWorkshopLink(option_id: $option_id, url: $url)
      }
    `;
    const variables = { option_id: "option_id1", url: "https://newurl" };
    const res = await query({ query: SET_WORKSHOP_LINK, variables });
    expect(res.errors).toBeUndefined();
    expect(res.data.setWorkshopLink).toBe(true);
  });
  test("saves workshop attendance", async () => {
    const SAVE_ATTENDANCE = gql`
      mutation saveAttendance(
        $input: [AttendingStudent!]!
        $option_id: ID!
        $teacher_id: ID!
      ) {
        saveWorkshopsAttendance(
          input: $input
          option_id: $option_id
          teacher_id: $teacher_id
        )
      }
    `;

    const variables: MutationSaveWorkshopsAttendanceArgs = {
      input: [
        {
          codigo: "1234567890",
          nombre: "Benito Antonio",
          apellido_paterno: "Martinez",
          apellido_materno: "Ocasio",
          nivel: "4",
          grupo: "E4-1",
          workshop: "conversation",
          teacher: "Fulanito",
          attended: true
        }
      ],
      option_id: "1",
      teacher_id: "1"
    };

    const res = await query({ query: SAVE_ATTENDANCE, variables });

    expect(res.errors).toBeUndefined();
    expect(databaseAPI.db.none).toHaveBeenCalled();
    expect(workshopsSheetsAPI.append).toHaveBeenCalled();
    expect(res.data.saveWorkshopsAttendance).toBe(true);
  });
});

describe("teacher links page", () => {
  test("names and ids", async () => {
    const GET_TEACHERS_PAGE = gql`
      query getTeacherLinks {
        teachers {
          id
          name
        }
      }
    `;
    const res = await query({
      query: GET_TEACHERS_PAGE
    });
    expect(res.errors).toBeUndefined();
    expect(res.data).toMatchSnapshot();
  });

  test("dashboard", async () => {
    const QUERY_GET_TEACHERS_DASHBOARD = gql`
      query TeacherDashboard {
        teacher(id: "4") {
          id
          name
          options {
            id
            workshop_name
            time
            day
            reservations {
              codigo
              nombre
            }
          }
        }
      }
    `;
    const res = await query({ query: QUERY_GET_TEACHERS_DASHBOARD });
    expect(res.errors).toBeUndefined();
    expect(res.data).toMatchSnapshot();
  });
});
