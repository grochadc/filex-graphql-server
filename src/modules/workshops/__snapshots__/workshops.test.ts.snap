// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gets all workshops 1`] = `
Object {
  "workshops": Array [
    Object {
      "description": "This is the conversation workshop description.",
      "levels": Array [
        "1",
        "2",
        "3",
        "4",
        "5",
        "6",
      ],
      "name": "Conversation",
      "options": Array [
        Object {
          "available": true,
          "day": "Lunes",
          "id": "1",
          "teacher_name": "Fulanito",
          "time": "1300 - 14:00",
          "url": "http://url1",
          "workshop_name": "Conversation",
          "zoom_id": null,
        },
        Object {
          "available": true,
          "day": "Martes",
          "id": "2",
          "teacher_name": "Fulanito",
          "time": "13:00 - 14:00",
          "url": "http://url2",
          "workshop_name": "Conversation",
          "zoom_id": null,
        },
      ],
    },
  ],
}
`;

exports[`reservations makes a reservation 1`] = `
Array [
  1,
  1,
  null,
]
`;

exports[`reservations makes a reservation 2`] = `
Object {
  "makeWorkshopReservation": Object {
    "day": "Lunes",
    "id": "1",
    "teacher_id": "teacher_id1",
    "teacher_name": "Fulanito",
    "time": "1300 - 14:00",
    "url": "http://url1",
    "workshop_id": "workshop_id1",
    "workshop_name": "Conversation",
    "zoom_id": null,
  },
}
`;

exports[`reservations makes a tutorials reservation 1`] = `
Array [
  1,
  1,
  "negative questions",
]
`;

exports[`reservations makes a tutorials reservation 2`] = `
Object {
  "makeWorkshopReservation": Object {
    "day": "Lunes",
    "id": "1",
    "teacher_id": "teacher_id1",
    "teacher_name": "Fulanito",
    "time": "1300 - 14:00",
    "url": "http://url1",
    "workshop_id": "workshop_id1",
    "workshop_name": "Conversation",
    "zoom_id": null,
  },
}
`;

exports[`student data already registered 1`] = `
Object {
  "student": Object {
    "nombre": "Alberto",
    "reservation": Object {
      "day": "Lunes",
      "teacher_name": "Fulanito",
      "time": "1300 - 14:00",
      "url": "http://url1",
      "workshop_name": "Conversation",
    },
  },
}
`;

exports[`student data normal not registered 1`] = `
Object {
  "student": Object {
    "nombre": "Benito Antonio",
    "reservation": null,
  },
}
`;

exports[`teacher dashboard no reservations 1`] = `
Object {
  "teacher": Object {
    "id": "3",
    "name": "Sutanito",
    "options": Array [
      Object {
        "day": "Lunes",
        "id": "5",
        "reservations": null,
        "time": "13:00 - 14:00",
        "url": "https://meet.google.com/sutanitolink1",
      },
      Object {
        "day": "Martes",
        "id": "6",
        "reservations": null,
        "time": "13:00 - 14:00",
        "url": "https://meet.google.com/sutanitolink2",
      },
      Object {
        "day": "Viernes",
        "id": "21",
        "reservations": null,
        "time": "13:00 - 14:00",
        "url": "https://meet.google.com/sutanitolink3",
      },
    ],
  },
}
`;

exports[`teacher dashboard normal 1`] = `
Object {
  "teacher": Object {
    "id": "4",
    "name": "Gonzalo",
    "options": Array [
      Object {
        "day": "Miercoles",
        "id": "7",
        "reservations": Array [
          Object {
            "apellido_materno": "Ocasio",
            "apellido_paterno": "Martinez",
            "codigo": "1234509876",
            "grupo": "E4-1",
            "nivel": "4",
            "nombre": "Benito Antonio",
          },
        ],
        "time": "13:00 - 14:00",
        "url": "https://meet.google.com/ukv-avjt-qiy",
      },
      Object {
        "day": "Jueves",
        "id": "8",
        "reservations": Array [
          Object {
            "apellido_materno": "Ocasio",
            "apellido_paterno": "Martinez",
            "codigo": "1234509876",
            "grupo": "E4-1",
            "nivel": "4",
            "nombre": "Benito Antonio",
          },
        ],
        "time": "13:00 - 14:00",
        "url": "https://meet.google.com/ukv-avjt-qiy",
      },
      Object {
        "day": "Martes",
        "id": "22",
        "reservations": Array [
          Object {
            "apellido_materno": "Ocasio",
            "apellido_paterno": "Martinez",
            "codigo": "1234509876",
            "grupo": "E4-1",
            "nivel": "4",
            "nombre": "Benito Antonio",
          },
        ],
        "time": "13:00 - 14:00",
        "url": "https://meet.google.com/kux-jhii-eyi",
      },
    ],
  },
}
`;

exports[`teacher links page dashboard 1`] = `
Object {
  "teacher": Object {
    "id": "4",
    "name": "Gonzalo",
    "options": Array [
      Object {
        "day": "Miercoles",
        "id": "7",
        "reservations": Array [
          Object {
            "codigo": "1234509876",
            "nombre": "Benito Antonio",
          },
        ],
        "time": "13:00 - 14:00",
        "workshop_name": "Conversation",
      },
      Object {
        "day": "Jueves",
        "id": "8",
        "reservations": Array [
          Object {
            "codigo": "1234509876",
            "nombre": "Benito Antonio",
          },
        ],
        "time": "13:00 - 14:00",
        "workshop_name": "Conversation",
      },
      Object {
        "day": "Martes",
        "id": "22",
        "reservations": Array [
          Object {
            "codigo": "1234509876",
            "nombre": "Benito Antonio",
          },
        ],
        "time": "13:00 - 14:00",
        "workshop_name": "Tutoring",
      },
    ],
  },
}
`;

exports[`teacher links page names and ids 1`] = `
Object {
  "teachers": Array [
    Object {
      "id": "1",
      "name": "Fulanito",
    },
    Object {
      "id": "2",
      "name": "Menganito",
    },
  ],
}
`;
